src/java/org/apache/cassandra/service/consensus/migration/MigrationState.java

```diff
@@ -0,0 +1,218 @@
+import static org.apache.cassandra.utils.CollectionSerializers.serializedMapSize;
+
+// TODO this will mostly go away once we can move TableMigrationState into the table schema
+public class MigrationState implements MetadataValue<MigrationState>
+{
+    public static MigrationState EMPTY = new MigrationState(Epoch.EMPTY, ImmutableMap.of());
+    @Nonnull
+    public final Map<TableId, TableMigrationState> tableStates;
+
+    public final Epoch lastModified;
+
+    public MigrationState(@Nonnull Epoch lastModified, @Nonnull Map<TableId, TableMigrationState> tableStates)
+    {
+        checkNotNull(tableStates, "tableStates is null");
+        checkNotNull(lastModified, "lastModified is null");
+        this.lastModified = lastModified;
+        this.tableStates = ImmutableMap.copyOf(tableStates);
+    }
+
+    public Map<String, Object> toMap(@Nullable Set<String> keyspaceNames, @Nullable Set<String> tableNames)
+    {
+        return ImmutableMap.of("lastModifiedEpoch", lastModified.getEpoch(),
+                               "tableStates", tableStatesAsMaps(keyspaceNames, tableNames),
+                               "version", PojoToString.CURRENT_VERSION);
+    }
+
+    private List<Map<String, Object>> tableStatesAsMaps(@Nullable Set<String> keyspaceNames,
+                                                        @Nullable Set<String> tableNames)
+    {
+        ImmutableList.Builder<Map<String, Object>> builder = ImmutableList.builder();
+        for (TableMigrationState tms : tableStates.values())
+        {
+            if (keyspaceNames != null && !keyspaceNames.contains(tms.keyspaceName))
+                continue;
+            if (tableNames != null && !tableNames.contains(tms.tableName))
+                continue;
+            builder.add(tms.toMap());
+        }
+        return builder.build();
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        MigrationState that = (MigrationState) o;
+        return tableStates.equals(that.tableStates);
+    }
+
+    private static void withRangesMigrating(Map<TableId, TableMigrationState> current, ImmutableMap.Builder<TableId, TableMigrationState> next, TableMetadata metadata, List<Range<Token>> ranges, boolean overwrite)
+    {
+        TableMigrationState tableState;
+        ConsensusMigrationTarget target = ConsensusMigrationTarget.fromTransactionalMode(metadata.params.transactionalMode);
+        if (!overwrite && current.containsKey(metadata.id))
+        {
+            tableState = current.get(metadata.id).withRangesMigrating(ranges, target);
+        }
+        else
+        {
+            tableState = new TableMigrationState(metadata.keyspace, metadata.name, metadata.id, target, ImmutableSet.of(), ImmutableMap.of(Epoch.EMPTY, ranges));
+        }
+        next.put(metadata.id, tableState);
```

Comment:
Does this actually work? I thought `ImmutableMap.Builder` throws an exception on duplicate keys when you call `build()`.

There is apparently a `buildKeepingLast` method which should work?